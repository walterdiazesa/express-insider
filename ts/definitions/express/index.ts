export enum HandlerType {
  ROUTE = "Route",
  MIDDLEWARE = "Middleware",
  ROUTE_MIDDLEWARE = "Route Middleware",
}

export * from './objects';
export * from './verbs';