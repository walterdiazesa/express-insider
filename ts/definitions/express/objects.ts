import { TrailOptions } from "..";
import { HandlerType, Method } from "../../../ts";

export type RouteMatcher = {
  route: TrailOptions['ignoreRoutes'][number]['route'],
  method: Lowercase<TrailOptions['ignoreRoutes'][number]['method']>,
};

export type Route = {
  path: `/${string}`;
  stack: RouteStack[];
  methods: Record<Method, boolean>;
};

type Stack<T extends HandlerType | undefined = undefined> = {
  handle: Function & { stack?: (StackItem | StackItem<HandlerType.ROUTE>)[] };
  name: string;
  params?: undefined;
  path: T extends HandlerType.ROUTE ? string : unknown;
  keys: Record<string, string | boolean | number>[];
  regexp: RegExp;
};

export type StackItem<T extends HandlerType | undefined = undefined> = Stack<T> & {
  route: T extends undefined ? undefined | Route : T extends HandlerType.ROUTE ? Route : undefined;
  // Config params
  ignore: T extends HandlerType.ROUTE ? undefined | true : never;
  showResponse: T extends HandlerType.ROUTE ? undefined | true : never;
  showRequestedURL: T extends HandlerType.ROUTE ? undefined | true : never;
}

export type RouteStack<T extends Method = Method> = Stack & {
  method: T;
};