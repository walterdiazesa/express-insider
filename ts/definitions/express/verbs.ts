export type Method = "get" | "post" | "options" | "all" | "patch" | "delete" | "put" | "head" | "connect" | "trace";