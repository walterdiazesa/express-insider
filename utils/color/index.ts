import { COLOR } from "../../constants";

export const getStatusCodeColor = (statusCode: number) => (statusCode >= 200 && statusCode <= 299 ? COLOR.fgGreen : statusCode >= 400 ? COLOR.fgRed : COLOR.fgWhite);