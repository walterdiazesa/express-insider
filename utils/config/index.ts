import { Find, Method, Route, RouteMatcher, TrailOptions } from "../../ts";
import { Config, ConfigKeys } from "../../ts/definitions/app/config";

// const digest = (property: LooseAutocomplete<ConfigKeys[number]>): keyof ConfigMapper => {
export const digest = <T extends ConfigKeys[number]>(property: T): Find<ConfigKeys, T> => {
  switch (property) {
    // @ts-ignore
    case "trailId": return 0;
    // @ts-ignore
    case "logger": return 1;
    // @ts-ignore
    case "ignoreMiddlewares": return 2;
    // @ts-ignore
    case "ignoreRoutes": return 3;
    // @ts-ignore
    case "showRequestedURL": return 4;
    // @ts-ignore
    case "showResponse": return 5;
    // @ts-ignore
    case "showRSS": return 6;
    // @ts-ignore
    case "showColors": return 7;
    // @ts-ignore
    case "report": return 8;
    // @ts-ignore
    case "timingFormatter": return 9;
    // @ts-ignore
    case "initialImmutableMiddlewares": return 10;
    // @ts-ignore
    case "logStrategy": return 11;
    // @ts-ignore
    case "delayMs": return 12;
    // @ts-ignore
    case "skip": return 13;
    /* istanbul ignore next */
    default:
      // TypeScript will raise an error if the default case is reached.
      const _: never = property;
      throw new Error(`Property "${String(property)}" not supported.`)
  }
}

export const generateRouteMatcherGroup = (matcher: Config['ignoreRoutes'] | boolean) => {
  if (typeof matcher === 'boolean') return matcher;
  const matchGroup: (`/${string}` | RouteMatcher['method'][])[] = [] as any;
  for (let i = 0; i < matcher.length; i++) {
    const matchItem = matcher[i];
    const enlisted = matchGroup.findIndex((route) => route === matchItem.route);
    const method = matchItem.method.toLowerCase() as RouteMatcher['method'];
    if (enlisted !== -1) {
      const enlistedItem = matchGroup[enlisted + 1] as RouteMatcher['method'][];
      if (!enlistedItem.includes(method)) enlistedItem.push(method)
    }
    else matchGroup.push(matchItem.route, [method])
  }
  return matchGroup;
}