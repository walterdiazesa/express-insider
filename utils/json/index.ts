import { COLOR, TYPE_TO_COLOR, JSONTagPad } from "../../constants";

const valueFormatter = (value: any) => {
  const formatted =  typeof value === "string" ? `${COLOR.fgGreen}"${value}"` : `${TYPE_TO_COLOR[value === false || value === null ? "red" : typeof value]}${value}`
  return `${COLOR.reset}${formatted}${COLOR.reset}`
};
export const colorizedJSON = (nextLinePad: string, value: any, objMultiline = true, firstMultiline = true): string => {
  if (typeof value !== "object" || value === null) return valueFormatter(value)
  if (Array.isArray(value)) return `[\n${nextLinePad}${JSONTagPad}${value
      .map(item => colorizedJSON(`${nextLinePad}${JSONTagPad}`, item, objMultiline)).join(`,\n${nextLinePad}${JSONTagPad}`)}\n${nextLinePad}]`;
  const objEntries = Object.entries(value);
  if (!objEntries.length || !objEntries.some(([key, value]) => (typeof value === "object" ? Object.keys(value).length : true))) return "";
  if (!firstMultiline) return `{ ${objEntries.filter(([qK, qV]) => qV !== undefined).map(([qK, qV]) => `${COLOR.fgYellow}"${qK}": ${colorizedJSON(nextLinePad, qV, objMultiline)}`).join(", ")} }`
  return `{${objEntries
    .reduce((acc, [key, value]) => {
      if (value === undefined) return acc;
      const elements = `${acc},\n${nextLinePad}${JSONTagPad}${COLOR.fgYellow}${key}: ${COLOR.reset}`
      if (typeof value !== "object" || value === null) return `${elements}${valueFormatter(value)}`;
      const valueKeys = Object.keys(value);
      if (!valueKeys.length) return acc;
      return `${elements}${colorizedJSON(`${nextLinePad}${JSONTagPad}`, value, objMultiline, objMultiline)}`;
    }, "")
    .slice(1)}\n${nextLinePad}}`;
};