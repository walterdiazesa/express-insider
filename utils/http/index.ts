import { Response } from "express";
import { HTTP_STATUS_MESSAGE_TO_CODE } from "../../constants";

/**
  * __[ODD-1]__: When using express@4.9.0 (unknown about other versions under 4.10.0) the test "trail on custom configuration â€º Should handle [CASE 12] with POST /conditional-route"
  * fails, even that the route return statusCode 200 as expected, the trail output shows 404, when removing the next call in the third route everything works as expected,
  * what is happenning is that express treats undefined next routes different in 4.9.0, so as no afterwards middleware is found the app crashes setting the statusCode as 404, so
  * the bandage fix to support 4.9.0 is to redefine the possible outcome for 404 cases if res.statusMessage is still defined as "OK"
  * 
  * Current state: FIXED!, we could return => res.statusCode directly, but it's a good redundant condition
  */
export const getStatusCode = (res: Response): typeof HTTP_STATUS_MESSAGE_TO_CODE[keyof typeof HTTP_STATUS_MESSAGE_TO_CODE] =>
  res.statusCode === 404 && res.statusMessage !== 'Not found' ? HTTP_STATUS_MESSAGE_TO_CODE[res.statusMessage.toLowerCase().replaceAll(' ', '')] : res.statusCode;