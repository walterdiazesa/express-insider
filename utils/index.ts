import { Express } from 'express';
import { getCfg } from '../config';
import { HandlerType, Route, RouteMatcher, StackItem } from '../ts';
import { generateRouteMatcherGroup } from './config';

const config = getCfg();

export const getRuntime = () => {
  const runtime = (process as any).isBun || process.versions.bun ? 'bun' : 'node'
  const { engines } = require('../package.json');
  return {
    runtime,
    version: process.versions[runtime],
    trailCompatibility: engines[runtime]
  }
}

const sanitizePackageVersion = (version: string) => version.match(/[0-9.]+/).shift() as `${number}.${number}.${number}`;

export const isRuntimeCompatible = (runtime: ReturnType<typeof getRuntime>) => sanitizePackageVersion(runtime.version).localeCompare(sanitizePackageVersion(runtime.trailCompatibility), undefined, { numeric: true, sensitivity: 'base' }) >= 0

export const getRSS = (): '' | ` [${number} bytes]` => {
  if (!config[6]) return ''
  // BUN: memoryUsage on process since v0.6.14
  let rss = process.memoryUsage?.().rss;
  /* istanbul ignore next */ 
  if (rss === undefined) {
    try {
      // BUN: bun:jsc available since v0.5.1
      const { memoryUsage } = require('bun:jsc');
      rss = memoryUsage().current
    } catch (e) {}
  }
  return ` [${rss} bytes]`
}

export const isStackItemRoute = (stackItem: StackItem | StackItem<HandlerType.ROUTE>): stackItem is StackItem<HandlerType.ROUTE> => typeof stackItem.route === 'object';

export const formatAnonymousRoute = (idx: number): `<anonymous (${number})>` => `<anonymous (${idx})>`;

/**
 * __[ODD-3]__: Under 4.0 ```app.stack``` is defined, after is ```app._router.stack```
 */
export const getStack = (app: Express): (StackItem | StackItem<HandlerType.ROUTE>)[] => app.stack || app._router.stack

/*
  PAPER: Let's start by saying that the timing difference between the worst possible matcher (a simple naive implementation
  without caching or transforming the original RouteMatcher given by the consumer) and this approach of grouping and storing
  the RouteMatchers in pairs inside an array is insignificant, not only in real-case scenarios but also relatively speaking.

  However, if I can either mantain the same performance (in the worst case) or improve it just a little bit without causing
  any issues in the ecosystem (the app itself or the debugging process for other developers, though in this case, I'm the
  only one :p), why not?

  My first approach was caching results by grouping a Map inside a Map (for caching, this was the faster implementation). I
  also tested objects, storing keys instead of references, and using the route as the identifier for the outer Map and the
  matchGroup for the inner Map (and vice versa), etc.
  
  But there were many variables that significantly affected the performance results:
    - The order of the route matcher requested versus the route that is actually being requested
      (User requesting GET / --> Match:
          [{ route: '/', method: 'post' }, { route: '/todo', method: 'get' }, { route: '/', method: 'get' }])
          is different than
          [{ route: '/', method: 'get' }, { route: '/todo', method: 'get' }, { route: '/', method: 'post' }])
    - And even worst, the size of the RouteMatcher set, at which point exactly is caching worth it? 3 items? 9 items? 20 items?
      keep in mind that the previous constraint (The order of the route matcher) is also affecting the second point both
      separately and as a whole
          Naive faster: [{ route: '/', method: 'post' }, { route: '/todo', method: 'get' }, { route: '/', method: 'get' }]
          Map faster?: [{ route: '/', method: 'post' }, (3 items more...), { route: '/', method: 'get' }]

  Then, I thought of transforming the original RouteMatcher into an array because, as I already mentioned in Config, using
  arrays is faster than any other type of object (even though arrays are also "objects").

  So, I came up with an array of arrays ([['/', ['get', 'post']], ['/todo', ['put']]]) which worked better than the previous
  approaches in all the cases.
  
  But then I thought, "Mmmmm, maybe using a top-level array of pairs with only the methods encapsuled in an array is faster."
  Thus, instead of [['/', ['get', 'post']], ['/todo', ['put']]]
  I used ['/', ['get', 'post'], '/todo', ['put']] and my hypothesis was correct â€“ which is the approach you're seeing now.
*/
export const isRouteMatching = (route: Route, matchGroup: ReturnType<typeof generateRouteMatcherGroup>) => {
  if (!matchGroup || typeof matchGroup === 'boolean') return !!matchGroup;
  const path = route.path;
  for (let i = 0; i < matchGroup.length; i+=2) {
    const matchItemPath = matchGroup[i] as Route['path'];
    if (matchItemPath !== path) continue;
    const matchItemMethods = matchGroup[i + 1] as RouteMatcher['method'][];
    for (let j = 0; j < matchItemMethods.length; j++) {
      const method = matchItemMethods[j];
      if (method === 'any' || route.methods[method]) {
        return true;
      }
    }
  }
  return false
}

export * from './report'
export * from './json'
export * from './color'
export * from './logger'
export * from './http'
export * from './config'