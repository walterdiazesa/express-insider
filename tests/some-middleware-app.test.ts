import request from 'supertest'
import express from 'express'
import { trail } from '../index';
import { awaitConsoleLog } from './utils';

export const someMiddlewareApp = express();

someMiddlewareApp.use(function mockBeforeMiddleware(req, res, next) {
  next();
})

someMiddlewareApp.get('/', (req, res) => {
  res.send({ working: true })
})

someMiddlewareApp.get('/some-route', async function (req, res, next) {
  next();
}, async function mockAfterSomeRouteRouteMiddleware(req, res, next) {
  res.send({ working: true });
}, async function neverRouteMiddleware(req, res) {})

trail(someMiddlewareApp, {
  trailId: () => 'test-id',
  timingFormatter: (elapsed) => 0,
  ignoreMiddlewares: ['expressInit']
});

describe('trail with no middleware output', () => {
  it('Should show handler message for GET / route', async () => {
    const log = jest.spyOn(global.console, 'log');
    await request(someMiddlewareApp).get("/");
    await awaitConsoleLog(0, log.mock.calls)
    expect(log.mock.calls).toMatchSnapshot();
  })
  it('Should show handler message for GET /some-route route', async () => {
    const log = jest.spyOn(global.console, 'log');
    await request(someMiddlewareApp).get("/some-route");
    await awaitConsoleLog(0, log.mock.calls)
    expect(log.mock.calls).toMatchSnapshot();
  })
})