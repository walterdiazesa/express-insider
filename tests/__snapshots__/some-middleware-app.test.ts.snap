// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`trail with no middleware output Should show handler message for GET / route 1`] = `
[
  [
    "[test-id] [32mGET / [0m[1mstart[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mquery [33m0 ms[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mmockBeforeMiddleware [33m0 ms[0m",
  ],
  [
    "[test-id] [35mRoute [32mGET / [33m0 ms [32m200 OK[0m",
  ],
  [
    "[test-id] [32mGET / [0m[1mfinish[0m, elapsed time since begin: [33m0 ms[0m",
  ],
]
`;

exports[`trail with no middleware output Should show handler message for GET /some-route route 1`] = `
[
  [
    "[test-id] [32mGET /some-route [0m[1mstart[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mquery [33m0 ms[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mmockBeforeMiddleware [33m0 ms[0m",
  ],
  [
    "[test-id] [35mRoute [32mGET /some-route[0m",
  ],
  [
    "[test-id]                      [32mmockAfterSomeRouteRouteMiddleware[0m",
  ],
  [
    "[test-id]                                                       (send response) [33m0 ms[0m",
  ],
  [
    "[test-id]                                                       [32m200 OK[0m, total elapsed time: [33m0 ms[0m",
  ],
  [
    "[test-id]                                                       (cleanup) [33m0 ms[0m",
  ],
  [
    "[test-id]                                                       (total) [33m0 ms[0m",
  ],
  [
    "[test-id]                      [32m<anonymous (0)> [33m0 ms[0m",
  ],
  [
    "[test-id] [32mGET /some-route [0m[1mfinish[0m, elapsed time since begin: [33m0 ms[0m",
  ],
]
`;
